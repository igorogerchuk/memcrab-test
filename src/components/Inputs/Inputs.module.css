.formWrapper {
    border-radius: 15px;
    border: 1px solid #999;
    box-shadow: 0px 2px 2px rgba(0, 0, 0, 0.5), 0px 0px 5px rgba(0, 0, 0, 0.25), 0px 0px 2px 4px rgba(0, 0, 0, 0.1);
    color: #444;
    font-weight: bold;
    text-shadow: 0px 1px 1px rgba(0, 0, 0, 0.3);
    background: linear-gradient(110deg, #eee 0%, #e3e3e3 10%, #fff 20%, #fff 25%, #f0f0f0 26%, #fff 28%, #ddd 55%, #eee 100%);
    padding: 30px;
    margin-bottom: 30px;
}

.label {
    font-size: 18px;
}

.input {
    border-radius: 5px;
    margin-right: 20px;
    outline: none;
    width: 80px;
    font-size: 18px;
    padding: 5px;
    color: #444;
    margin-bottom: 30px;
}

.submitButton {
    display: block;
    margin: 0 auto;
    font-size: 18px;
    outline: none;
    border-radius: 15px;
    border: 1px solid #999;
    box-shadow: 0px 2px 2px rgba(0, 0, 0, 0.5), 0px 0px 5px rgba(0, 0, 0, 0.25), 0px 0px 2px 4px rgba(0, 0, 0, 0.1);
    color: #444;
    font-weight: bold;
    text-shadow: 0px 1px 1px rgba(0, 0, 0, 0.3);
    background: linear-gradient(110deg, #eee 0%, #e3e3e3 10%, #fff 20%, #fff 25%, #f0f0f0 26%, #fff 28%, #ddd 55%, #eee 100%);
    padding: 30px;
}

.submitButton:hover {
    background: linear-gradient(110deg, #f90c04 0%, #ed413b 10%, #fff 20%, #fff 25%, #f0f0f0 26%, #fff 28%, #f53d37 55%, #f90c04 100%);
}

.submitButton:active {
    position: relative;
    top: 2px;
}