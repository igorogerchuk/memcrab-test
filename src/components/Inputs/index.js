// @flow
export { default } from "./Inputs";
