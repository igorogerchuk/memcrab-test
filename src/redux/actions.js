// @flow

import * as types from "./types";
import type { ArrayState, RowsState, CellsState, ParamsState } from "./types";

type _ExtractReturn<R, F: (...args: any[]) => R> = R;
type ExtractReturn<F> = _ExtractReturn<*, F>;

export const removeRow = (id: string, cellsIds: $ReadOnlyArray<string>) => ({
  type: types.REMOVE_ROW,
  payload: { id, cellsIds }
});

export const addRow = (
  array: ArrayState,
  row: RowsState,
  cells: CellsState
) => ({
  type: types.ADD_ROW,
  payload: { array, row, cells }
});

export const saveArray = (array: ArrayState) => ({
  type: types.SAVE_ARRAY,
  payload: { array }
});

export const saveRows = (rows: RowsState) => ({
  type: types.SAVE_ROWS,
  payload: { rows }
});

export const saveCells = (cells: CellsState) => ({
  type: types.SAVE_CELLS,
  payload: { cells }
});

export const increase = (id: string) => ({
  type: types.INCREASE,
  payload: { id }
});

export const saveParams = (params: ParamsState) => ({
  type: types.SAVE_PARAMS,
  payload: { params }
});

export type Action =
  | ExtractReturn<typeof removeRow>
  | ExtractReturn<typeof addRow>
  | ExtractReturn<typeof saveArray>
  | ExtractReturn<typeof saveRows>
  | ExtractReturn<typeof saveCells>
  | ExtractReturn<typeof saveParams>
  | ExtractReturn<typeof increase>;
