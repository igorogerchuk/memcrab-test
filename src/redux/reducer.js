import types from "./types";

const reducer = (state = { array: [] }, { type, payload }) => {
  const { array } = state;
  switch (type) {
    case types.ADD_LINE:
      return {
        array: array.concat(payload.newLine)
      };
    case types.REMOVE_LINE:
      return {
        array: array.filter(line => line !== payload.line)
      };
    case types.SAVE_ARRAY:
      return { array: payload.array };
    default:
      return state;
  }
};

export default reducer;
