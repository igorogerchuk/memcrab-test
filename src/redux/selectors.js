// @flow
import { createSelector } from "reselect";
import type { State } from "./types";

export const getArray = (state: State) => state.array;

export const getRows = (state: State) => state.rows;

export const getCells = (state: State) => state.cells;

export const getColumnsQty = (state: State) => state.params.n;

export const getIlluminatedQty = (state: State) => state.params.x;

export const getSumColumn = createSelector<State, *, *, *, *>(
  [getRows, getCells],
  (rows, cells) => {
    const cellsIds = Object.keys(rows).map((rowId: string) => rows[rowId]);
    const sumColumn = cellsIds.map((row: $ReadOnlyArray<string>) =>
      row.reduce((acc, id) => (acc += cells[id].amount), 0)
    );
    return sumColumn;
  }
);

export const getAverageRow = createSelector<State, *, *, *, *, *>(
  [getRows, getCells, getColumnsQty],
  (rows, cells, columnsQty) => {
    const cellsIds = Object.keys(rows).map((rowId: string) => rows[rowId]);
    const rowsQty = cellsIds.length;
    const averageRow = cellsIds.reduce(
      (acc: Array<number>, row: $ReadOnlyArray<string>) =>
        row.map((id, index) => (acc[index] += cells[id].amount / rowsQty)),
      new Array(columnsQty).fill(0)
    );
    return averageRow;
  }
);
