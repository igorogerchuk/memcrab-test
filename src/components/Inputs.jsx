import React, { Component } from "react";
import { connect } from "react-redux";
import * as actions from "../redux/actions";
import createRandomArray from "../services/randomArray";

class Inputs extends Component {
  state = { m: "", n: "", x: "" };

  inputHandler = e => {
    const { id, value } = e.target;
    switch (id) {
      case "m":
        this.setState(state => ({ ...state, m: value }));
        break;
      case "n":
        this.setState(state => ({ ...state, n: value }));
        break;
      case "x":
        this.setState(state => ({ ...state, x: value }));
        break;
      default:
        break;
    }
  };

  submitHandler = e => {
    e.preventDefault();
    const { m, n, x } = this.state;
    const { onSaveArray } = this.props; //, onSaveParams
    if (m <= 0 || n <= 0 || x < 0) {
      alert("array parameters must be more then 0");
      return;
    }
    onSaveArray(createRandomArray(m, n));
    // onSaveParams({ n, x });

    this.setState({ m: "", n: "", x: "" });
  };

  render() {
    const { m, n, x } = this.state;
    return (
      <div>
        <p>Enter table parameters</p>
        <form onSubmit={this.submitHandler}>
          <label>
            Lines quantity:{" "}
            <input
              type="number"
              name="lines"
              id="m"
              onChange={this.inputHandler}
              value={m}
              placeholder="0"
            />
          </label>
          <label>
            Columns quantity:
            <input
              type="number"
              name="columns"
              id="n"
              onChange={this.inputHandler}
              value={n}
              placeholder="0"
            />
          </label>
          <label>
            Illuminated cells quantity:
            <input
              type="number"
              name="illuminate"
              id="x"
              onChange={this.inputHandler}
              value={x}
              placeholder="0"
            />
          </label>
          <input type="submit" value="Generate" />
        </form>
      </div>
    );
  }
}

const mapDispatchToProps = dispatch => {
  return {
    onSaveArray: array => dispatch(actions.saveArray(array))
    // onSaveParams: params => dispatch(actions.saveParams(params))
  };
};

export default connect(null, mapDispatchToProps)(Inputs);
