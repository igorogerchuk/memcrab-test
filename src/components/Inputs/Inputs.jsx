// @flow
import React, { useState } from "react";
import { useDispatch, batch } from "react-redux";
import * as actions from "../../redux/actions";
import type { Action } from "../../redux/actions";
import type { Dispatch } from "redux";
import createRandomArray from "../../services/randomArray";
import "./Inputs.module.css";

const Inputs = () => {
  const dispatch = useDispatch<Dispatch<Action>>();
  const [params, setParams] = useState({ m: "", n: "", x: "" });
  const { m, n, x } = params;

  const inputHandler = (e: SyntheticInputEvent<HTMLInputElement>) => {
    const { id, value } = e.target;
    setParams({ ...params, [id]: value });
  };

  const submitHandler = (e: SyntheticInputEvent<HTMLInputElement>) => {
    e.preventDefault();
    const rowsQty = +m;
    const columnsQty = +n;
    const illuminatedQty = +x;
    if (rowsQty <= 0 || columnsQty <= 0 || illuminatedQty < 0) {
      alert("array parameters must be more then 0");
      return;
    }
    batch(() => {
      dispatch(actions.saveParams({ n: columnsQty, x: illuminatedQty }));

      const { array, rows, cells } = createRandomArray(rowsQty, columnsQty);
      dispatch(actions.saveArray(array));
      dispatch(actions.saveRows(rows));
      dispatch(actions.saveCells(cells));
    });
    setParams({ m: "", n: "", x: "" });
  };

  return (
    <div styleName="formWrapper">
      <h1 styleName="formTitle">Enter table parameters</h1>
      <form styleName="form" onSubmit={submitHandler} id="form">
        <div styleName="inputWrapper">
          <label htmlFor="m" styleName="label">
            Rows quantity:
          </label>
          <input
            styleName="input"
            type="number"
            name="rows"
            id="m"
            onChange={inputHandler}
            value={m}
            placeholder="0"
          />
        </div>
        <div styleName="inputWrapper">
          <label htmlFor="n" styleName="label">
            Columns quantity:
          </label>
          <input
            styleName="input"
            type="number"
            name="columns"
            id="n"
            onChange={inputHandler}
            value={n}
            placeholder="0"
          />
        </div>
        <div styleName="inputWrapper">
          <label htmlFor="m" styleName="label">
            Illuminated cells quantity:
          </label>
          <input
            styleName="input"
            type="number"
            name="illuminate"
            id="x"
            onChange={inputHandler}
            value={x}
            placeholder="0"
          />
        </div>
        <div>
          <input styleName="submitButton" type="submit" value="Generate" />
        </div>
      </form>
    </div>
  );
};

export default Inputs;
