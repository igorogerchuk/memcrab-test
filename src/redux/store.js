// @flow
import { createStore } from "redux";
import reducer from "./reducer";
import type { State } from "./types";
import type { Dispatch } from "redux";
import type { Action } from "./actions";

const store = createStore<State, Action, Dispatch<Action>>(
  reducer,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__()
);
export default store;
