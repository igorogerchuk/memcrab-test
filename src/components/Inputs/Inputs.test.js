import React from "react";
import Inputs from "./Inputs";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { createStore } from "redux";
import reducer from "../../redux/reducer";

describe("testing Input methods", () => {
  const store = createStore(reducer);
  let component;
  renderer.act(() => {
    component = renderer.create(
      <Provider store={store}>
        <Inputs />
      </Provider>
    );
  });
  const instance = component.root;

  it("should render correctly after some number was typed into form input", () => {
    const id = "m";
    const e = { target: { id, value: 5 } };
    const input = instance.findByProps({ id });
    renderer.act(() => input.props.onChange(e));
    const snap = component.toJSON();
    expect(snap).toMatchSnapshot();
  });

  it("should render correctly after form submit", () => {
    const id = "n";
    const e = { target: { id, value: 4 } };
    const input = instance.findByProps({ id });
    renderer.act(() => input.props.onChange(e));
    const form = instance.findByProps({ id: "form" });
    const se = { preventDefault: jest.fn() };
    renderer.act(() => form.props.onSubmit(se));
    const snap = component.toJSON();
    expect(snap).toMatchSnapshot();
  });

  it("should render correctly after form submit", () => {
    global.alert = jest.fn();

    const id = "x";
    const e = { target: { id, value: -3 } };
    const input = instance.findByProps({ id });
    renderer.act(() => input.props.onChange(e));
    const form = instance.findByProps({ id: "form" });
    const se = { preventDefault: jest.fn() };
    renderer.act(() => form.props.onSubmit(se));
    expect(global.alert).toHaveBeenCalledWith(
      "array parameters must be more then 0"
    );
  });
});
