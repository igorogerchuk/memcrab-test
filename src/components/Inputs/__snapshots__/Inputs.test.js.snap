// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing Input methods should render correctly after form submit 1`] = `
<div
  className="formWrapper___Xr_Ln"
>
  <h1
    className="formTitle___1Sczm"
  >
    Enter table parameters
  </h1>
  <form
    className="form___1t6C-"
    id="form"
    onSubmit={[Function]}
  >
    <div
      className="inputWrapper___1nGv5"
    >
      <label
        className="label___2sD9G"
        htmlFor="m"
      >
        Rows quantity:
      </label>
      <input
        className="input___3WuZL"
        id="m"
        name="rows"
        onChange={[Function]}
        placeholder="0"
        type="number"
        value=""
      />
    </div>
    <div
      className="inputWrapper___1nGv5"
    >
      <label
        className="label___2sD9G"
        htmlFor="n"
      >
        Columns quantity:
      </label>
      <input
        className="input___3WuZL"
        id="n"
        name="columns"
        onChange={[Function]}
        placeholder="0"
        type="number"
        value=""
      />
    </div>
    <div
      className="inputWrapper___1nGv5"
    >
      <label
        className="label___2sD9G"
        htmlFor="m"
      >
        Illuminated cells quantity:
      </label>
      <input
        className="input___3WuZL"
        id="x"
        name="illuminate"
        onChange={[Function]}
        placeholder="0"
        type="number"
        value=""
      />
    </div>
    <div>
      <input
        className="submitButton___1e-tC"
        type="submit"
        value="Generate"
      />
    </div>
  </form>
</div>
`;

exports[`testing Input methods should render correctly after some number was typed into form input 1`] = `
<div
  className="formWrapper___Xr_Ln"
>
  <h1
    className="formTitle___1Sczm"
  >
    Enter table parameters
  </h1>
  <form
    className="form___1t6C-"
    id="form"
    onSubmit={[Function]}
  >
    <div
      className="inputWrapper___1nGv5"
    >
      <label
        className="label___2sD9G"
        htmlFor="m"
      >
        Rows quantity:
      </label>
      <input
        className="input___3WuZL"
        id="m"
        name="rows"
        onChange={[Function]}
        placeholder="0"
        type="number"
        value={5}
      />
    </div>
    <div
      className="inputWrapper___1nGv5"
    >
      <label
        className="label___2sD9G"
        htmlFor="n"
      >
        Columns quantity:
      </label>
      <input
        className="input___3WuZL"
        id="n"
        name="columns"
        onChange={[Function]}
        placeholder="0"
        type="number"
        value=""
      />
    </div>
    <div
      className="inputWrapper___1nGv5"
    >
      <label
        className="label___2sD9G"
        htmlFor="m"
      >
        Illuminated cells quantity:
      </label>
      <input
        className="input___3WuZL"
        id="x"
        name="illuminate"
        onChange={[Function]}
        placeholder="0"
        type="number"
        value=""
      />
    </div>
    <div>
      <input
        className="submitButton___1e-tC"
        type="submit"
        value="Generate"
      />
    </div>
  </form>
</div>
`;
